stl/inc/functional

```diff
@@ -2171,48 +2171,48 @@ struct _Global_delete {
-        for (_FwdItPat _Mid2 = _First2;; ++_Mid1, (void) ++_Mid2) {
-            if (_Mid2 == _Last2) {
-                return {_First1, _Mid1};
-            }
-
-            if (_Mid1 == _Last1) {
-                return {_Last1, _Last1};
-            }
+_CONSTEXPR20 pair<_FwdItHaystack, _FwdItHaystack> _Search_pair_unchecked(
+    _FwdItHaystack _First1, _FwdItHaystack _Last1, _FwdItPat _First2, _FwdItPat _Last2, _Pred_eq& _Eq) {
+    if constexpr (_Is_random_iter_v<_FwdItHaystack> && _Is_random_iter_v<_FwdItPat>) {
+        // find first [_First2, _Last2) satisfying _Eq, random-access iterators
+        _Iter_diff_t<_FwdItHaystack> _Count1 = _Last1 - _First1;
+        _Iter_diff_t<_FwdItPat> _Count2      = _Last2 - _First2;
+
+        for (; _Count2 <= _Count1; ++_First1, (void) --_Count1) { // room for match, try it
+            _FwdItHaystack _Mid1 = _First1;
+            for (_FwdItPat _Mid2 = _First2;; ++_Mid1, (void) ++_Mid2) {
+                if (_Mid2 == _Last2) {
+                    return {_First1, _Mid1};
+                }
 
-            if (!_Eq(*_Mid1, *_Mid2)) {
-                break;
+                if (!_Eq(*_Mid1, *_Mid2)) {
+                    break;
+                }
             }
         }
-    }
-}
 
-template <class _FwdItHaystack, class _FwdItPat, class _Pred_eq>
-_CONSTEXPR20 pair<_FwdItHaystack, _FwdItHaystack> _Search_pair_unchecked(_FwdItHaystack _First1, _FwdItHaystack _Last1,
-    _FwdItPat _First2, _FwdItPat _Last2, _Pred_eq& _Eq, random_access_iterator_tag, random_access_iterator_tag) {
-    // find first [_First2, _Last2) satisfying _Eq, random-access iterators
-    _Iter_diff_t<_FwdItHaystack> _Count1 = _Last1 - _First1;
-    _Iter_diff_t<_FwdItPat> _Count2      = _Last2 - _First2;
-
-    for (; _Count2 <= _Count1; ++_First1, (void) --_Count1) { // room for match, try it
-        _FwdItHaystack _Mid1 = _First1;
-        for (_FwdItPat _Mid2 = _First2;; ++_Mid1, (void) ++_Mid2) {
-            if (_Mid2 == _Last2) {
-                return {_First1, _Mid1};
-            }
+        return {_Last1, _Last1};
+    } else if constexpr (_Is_fwd_iter_v<_FwdItHaystack> && _Is_fwd_iter_v<_FwdItPat>) {
+        // find first [_First2, _Last2) satisfying _Eq, arbitrary iterators
+        for (;; ++_First1) { // loop until match or end of a sequence
+            _FwdItHaystack _Mid1 = _First1;
+            for (_FwdItPat _Mid2 = _First2;; ++_Mid1, (void) ++_Mid2) {
+                if (_Mid2 == _Last2) {
+                    return {_First1, _Mid1};
+                }
 
-            if (!_Eq(*_Mid1, *_Mid2)) {
-                break;
+                if (_Mid1 == _Last1) {
+                    return {_Last1, _Last1};
+                }
+
+                if (!_Eq(*_Mid1, *_Mid2)) {
+                    break;
+                }
             }
         }
+    } else {
+        static_assert(_Always_false<_FwdItHaystack>, "Iterator parameters must be at least forward iterators");

```

Comment:
The parameters don't have concerete types - we mean the _arguments_ here. IMO, however, it would read better as simply "iterators".